### PR TITLE
Add type tracking and type info to jQuery model

### PR DESCRIPTION
**Pull request by asger-semmle**
- Moves jQuery definitions into the `JQuery` namespace,
- Uses type-tracking and type info for jQuery objects as well as the `$` function itself
- Makes jQuery attribute definitions work when the receiver is not an element creation, but a reference to an element.
- A few perf tweaks which help a little.

[Full eval on nightly](https://git.semmle.com/asger/dist-compare-reports/tree/jquery_1569339544812) shows that the analysis overall becomes a few percent slower in exchange for the added coverage. Unfortunately, the DIL changes are too large for me to isolate the cause of the slowdown.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Extend.qll [hunk 1]
```diff
@@ -47,7 +47,9 @@ private class ExtendCallWithFlag extends ExtendCall {
       name = "node.extend"
     )
     or
-    this = jquery().getAPropertyRead("extend").getACall()
+    // Match $.extend using the source of `$` only, as ExtendCall should not
```
[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/Extend.qll]: Monotonicity. `ClassNode` uses `ExtendCall` to recognise instance methods, which affects the call graph, which affects type tracking, which should therefore not affect `ExtendCall`. Even if the recursion wasn't negative I'd prefer not to create this dependency.

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 2]
```diff
@@ -318,4 +306,133 @@ module JQuery {
     name = "wrapAll" or
     name = "wrapInner"
   }
+
+  module DollarSource {
```
[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/jQuery.qll]: I'll add some comments

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 3]
```diff
@@ -318,4 +306,133 @@ module JQuery {
...
+    /**
+     * A data flow node that should be considered a source of jQuery objects.
+     */
+    abstract class Range extends DataFlow::Node { }
+
+    private class DefaultRange extends Range {
+      DefaultRange() {
+        this.asExpr() instanceof JQueryObjectInternal
+        or
+        hasUnderlyingType("JQuery")
```
[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/jQuery.qll]: I mean I suppose people can write whatever they want in jsdoc annotations, so in that sense, we could come across a lower-case `@param {jquery}`. But generally it needs to be the name of a variable that statically refers the class. What "statically refers to" means not very well standardized as far as I can tell.

It *could* become relevant for the combination of module systems and JSDoc annotations, like:
```javascript
let jquery = require('jquery');

/** @param {jquery} jq */
function f(jq) {}
```
The proper solution here would be to use the 2-argument version of `hasUnderlyingType`, namely `hasUnderlyingType("jquery", "")`. But in JSDoc we don't currently support this kind of name resolution, as I couldn't find any real-world examples to test on.

File: javascript/ql/src/semmle/javascript/frameworks/jQuery.qll [hunk 4]
```diff
@@ -394,6 +406,10 @@ module JQuery {
       this = dollar().getAMemberCall(name)
       or
       this = objectRef().getAMethodCall(name)
+      or
+      // Handle contributed JQuery objects that aren't source nodes (usually parameter uses)
```
[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/jQuery.qll]: > outside our control that we want to support in the name of backwards compatibility?

yes

File: javascript/ql/src/semmle/javascript/frameworks/AngularJS/AngularJSCore.qll [hunk 5]
```diff
@@ -652,15 +652,13 @@ private class JQLiteObject extends JQueryObject {
...
-    |
-      this = param.getAnInitialUse()
     )
     or
     exists(ServiceReference element |
       element.getName() = "$rootElement" or
       element.getName() = "$document"
     |
-      this = element.getAnAccess()
+      this = element.getAReference()
```
[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/AngularJS/AngularJSCore.qll]: This is already a `SourceNode`.

[the remark this answers is not in the document]
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/frameworks/AngularJS/AngularJSCore.qll]: It's already a `SourceNode`

**Comment by asger-semmle** [comment]
> The two pragma[inline] are a bit surprising - did you evaluate them in isolation in one of your many recent evalutions?

Yes but there's been so many evaluations I've lost track of which one was what. I can do another evaluation if needed.

**Comment by asger-semmle** [comment]
[Reverting the perf tweaks](https://git.semmle.com/asger/dist-compare-reports/tree/jquery-without-perf-tweaks_1569420323112) commit yields a minor speed up on smoke-test so I'll run a full eval on nightly without that commit to get to the bottom of this.

**Comment by asger-semmle** [comment]
[Evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/jquery-without-perf-tweaks_1570378039899) confirms it's better to remove the inlining, which I've done now.

**Comment by esben-semmle** [comment]
Sigh. Another test failure.

**Comment by esben-semmle** [comment]
Green, finally! @asger-semmle are we ready to merge, or are you evaluating the latest change?

**Comment by asger-semmle** [comment]
[Final evaluation](https://git.semmle.com/asger/dist-compare-reports/tree/jquery_1570585283927) looks good.

**Pull request merged**